### PR TITLE
Feature/oneteam-284

### PR DESCRIPTION
**Pull request by cvarjao**

### CODE REVIEW

File: src/api/model/vault/entry.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export default abstract class Entry {
```
Comment (mbystedt) [inline src/api/model/vault/entry.ts:1]: Can this be done as an interface instead?

File: src/api/service/axios-bitbucket-client.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  static parseUrl(url: string): RepositoryReference {
+    const regex = /https:\/\/(apps|bwa)\.nrs\.gov\.bc\.ca\/int\/stash\/projects\/(?<project>[^/]+)\/repos\/(?<repository>[^/\s]+)/gm
+    const m = regex.exec(url)
+    if (m === null) {
+      throw new Error(`Unable to parse BitBucket Url from ${url}`)
+    }
+    return {slug: m.groups?.repository as string, project: {key: m.groups?.project as string}}
+  }
+
+  constructor(client: AxiosInstance) {
```
Comment (mbystedt) [inline src/api/service/axios-bitbucket-client.ts:34]: constructor(readonly client: AxiosInstance) {}

Declaration above is not needed.

File: src/api/service/axios-factory.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import {AxiosBitBucketClient} from './axios-bitbucket-client'
+
+export class AxiosFactory {
+  static async jira(): Promise<AxiosJiraClient> {
+    const entry = await (await SecretManager.getInstance()).getEntry(SVC_IDIR_SPEC)
+    const idirUsername = (await entry.getProperty(SVC_IDIR_SPEC.fields.UPN.name)).getPlainText()
+    const idirPassword = (await entry.getProperty(SVC_IDIR_SPEC.fields.PASSWORD.name))
+
+    return new AxiosJiraClient(axios.create({
+      baseURL: 'https://bwa.nrs.gov.bc.ca/int/jira',
```
Comment (mbystedt) [inline src/api/service/axios-factory.ts:13]: May need to add the environment here as testing should be done on the test instance of jira.

File: src/api/service/axios-factory.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      timeout: 10000,
+      auth: {username: idirUsername, password: idirPassword.getPlainText()},
+    }))
+  }
+
+  static async bitBucket(): Promise<AxiosBitBucketClient> {
+    const entry = await (await SecretManager.getInstance()).getEntry(SVC_IDIR_SPEC)
+    const idirUsername = (await entry.getProperty(SVC_IDIR_SPEC.fields.UPN.name)).getPlainText()
+    const idirPassword = (await entry.getProperty(SVC_IDIR_SPEC.fields.PASSWORD.name))
```
Comment (mbystedt) [inline src/api/service/axios-factory.ts:22]: Duplicated code block.

File: src/api/service/axios-jira-client.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+type Issue = any
+
+export const FIELDS = Object.freeze({
+  ISSUE_TYPE: 'issuetype',
+})
+export class AxiosJiraClient {
+  readonly client: AxiosInstance
+
+  constructor(client: AxiosInstance) {
```
Comment (mbystedt) [inline src/api/service/axios-jira-client.ts:12]: constructor(readonly client: AxiosInstance) {}

Declaration above is not needed.

File: src/api/service/axios-jira-client.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+  public async getIssue(issueKey: string, params?: any): Promise<Issue> {
+    return this.client.get(`rest/api/2/issue/${issueKey}`, {params: params})
+    .then(response => {
+      return response.data
+    })
+  }
+
+  public getBranches(issueId: string, params?: any) {
+    return this.client.get('rest/dev-status/1.0/issue/detail', {params: Object.assign({}, params, {issueId: issueId, applicationType: 'stash', dataType: 'pullrequest'})}).then(response => {
```
Comment (mbystedt) [inline src/api/service/axios-jira-client.ts:24]: Prefer es6 object spread operator over Object.assign.

File: src/api/service/secret-manager.ts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+  /**
+   * Hide values from util.inspect, console.log, console.dir.
+   * @returns {string} Masked value
+   */
+  // [util.inspect.custom](): string {
+  //   return this.toString()
+  // }
+}
+
+class EntryAccessor {
```
Comment (mbystedt) [inline src/api/service/secret-manager.ts:41]: Multiple object in a single file. Please seperate.

File: src/api/service/secret-manager.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+          }
+          prompts.push({type: fieldSpec.type || 'input', name: fieldSpec.name, message: message})
+        }
+      }
+      const answers = await prompt(prompts)
+      for (const fieldName of Object.keys(answers)) {
+        svc[fieldName] = answers[fieldName]
+      }
+    }
+    if (!svc && !svc.name) {
```
Comment (mbystedt) [inline src/api/service/secret-manager.ts:131]: !svc?.name

File: src/commands/backlog/checkin.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
+import {GitBaseCommand} from '../../git-base'
+import {flags} from '@oclif/command'
+import {AxiosJiraClient} from '../../api/service/axios-jira-client'
+import {AxiosBitBucketClient} from '../../api/service/axios-bitbucket-client'
+
+export default class GitPush extends GitBaseCommand {
```
Comment (mbystedt) [inline src/commands/backlog/checkin.ts:6]: class name does not match filename.

File: src/commands/backlog/checkin.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  static description = 'Push local changes (commits) to the remote repository'
+
+  static flags = {
+    pr: flags.boolean({description: 'Create Pull-Request', default: true}),
+  }
+
+  async run() {
```
Comment (mbystedt) [inline src/commands/backlog/checkin.ts:13]: My eyes are defocusing looking at this. This should be split into multiple function calls to isolate and describe the steps going on here.

File: src/commands/backlog/checkout.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    return Promise.resolve(true)
+    .then(() => {
+      cli.action.start(message)
+      return promise
+    }).finally(() => {
+      cli.action.stop()
+    })
+  }
+
+  async run() {
```
Comment (mbystedt) [inline src/commands/backlog/checkout.ts:37]: Same as checkin.ts . This is not very readable and at a glance I can just say it does stuff. No idea if it doing the right thing.

File: src/git-base.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      levels: winston.config.cli.levels,
+      format: winston.format.combine(
+        winston.format.splat(),
+        winston.format.printf(info => `${info.level}: ${info.message}`)
+      ),
+      transports: [new winston.transports.Console()],
+    })
+  }
+
+  async _spawn(command: string, argsv?: readonly string[], options?: SpawnOptions): Promise<{status: number; stdout: string; stderr: string}> {
```
Comment (mbystedt) [inline src/git-base.ts:31]: This is not related to a GitBaseCommand. This is probably better in some kind of utility class.

File: src/git-base.ts [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+import {AxiosBitBucketClient, RepositoryReference} from './api/service/axios-bitbucket-client'
+import {AxiosJiraClient} from './api/service/axios-jira-client'
+import {AxiosFactory} from './api/service/axios-factory'
+import {SpawnOptions, spawn} from 'child_process'
+import * as winston from 'winston'
+import * as Config from '@oclif/config'
+
+export abstract class GitBaseCommand extends Command {
+  // eslint-disable-next-line no-useless-escape
+  static JIRA_ISSUE_KEY_REGEX = /(([^\/]+\/)+)?(?<issueKey>[^-]+-\d+)/gm;
```
Comment (mbystedt) [inline src/git-base.ts:11]: How is a Jira issue key related to git?

File: src/git-base.ts [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+      child.on('exit', status => {
+        resolve({status: status as number, stdout, stderr})
+      })
+    })
+  }
+
+  async createBranch(issue: any, repository: RepositoryReference, branchName: string, startPoint = 'master') {
+    const jira = this.jira as AxiosJiraClient
+    const bitBucket = this.bitBucket as AxiosBitBucketClient
+    const devDetails = (await jira.getBranches(issue.id))
```
Comment (mbystedt) [inline src/git-base.ts:52]: No contract that this.jira will be initialized. Null check on this.jira and this.bitbucket must be done.

File: test/commands/git/checkout.test.ts [hunk 15]
```diff
@@ -0,0 +1,17 @@
+import {expect, test} from '@oclif/test'
+
+describe.skip('git:checkout', () => {
```
Comment (mbystedt) [inline test/commands/git/checkout.test.ts:3]: Just my opinion. Lack of unit tests should be an auto fail of a PR.

File: src/git-base.ts [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    }
+    return {name: branchName, repository}
+  }
+
+  async createReleaseBranch(rfc: any, repository: RepositoryReference) {
+    const releaseBranch = `release/${rfc.key}`
+    return this.createBranch(rfc, repository, releaseBranch)
+  }
+
+  async init() {
```
Comment (mbystedt) [inline src/git-base.ts:69]: I would move this to under the constructor.

File: src/git-base.ts (hunk 14 above)
Comment (mbystedt) [reply src/git-base.ts:52]: Hmm... I see it now. I guess it works. The 'as AxiosJiraClient' makes me nervous. It shouldn't be necessary.

File: src/api/service/axios-bitbucket-client.ts (hunk 2 above)
Comment (cvarjao) [reply src/api/service/axios-bitbucket-client.ts:34]: @mbystedt,  I am not sure what you are asking, if I remove that constructor I get a type script error:
```
src/api/service/axios-bitbucket-client.ts:22:12 - error TS2564: Property 'client' has no initializer and is not definitely assigned in the constructor.

22   readonly client: AxiosInstance
              ~~~~~~
```
Can you please clarify your suggestion?

File: src/api/service/axios-jira-client.ts (hunk 5 above)
Comment (cvarjao) [reply src/api/service/axios-jira-client.ts:12]: TypeScript requires that block so it can properly validate/compile

File: src/api/service/secret-manager.ts (hunk 7 above)
Comment (cvarjao) [reply src/api/service/secret-manager.ts:41]: We will need to revisit this class and find a good abstraction/architecture

File: src/api/service/secret-manager.ts (hunk 8 above)
Comment (cvarjao) [reply src/api/service/secret-manager.ts:131]: addressed.

File: src/commands/backlog/checkin.ts (hunk 10 above)
Comment (cvarjao) [reply src/commands/backlog/checkin.ts:13]: For future, it may be easier to enforce lines of a function as part of eslint rule.

File: src/git-base.ts (hunk 12 above)
Comment (cvarjao) [reply src/git-base.ts:31]: separated out into `util/child-process.ts`

File: src/git-base.ts (hunk 13 above)
Comment (cvarjao) [reply src/git-base.ts:11]: moved to AxiosJiraClient

File: src/git-base.ts (hunk 16 above)
Comment (cvarjao) [reply src/git-base.ts:69]: moved after constructor

File: test/commands/git/checkout.test.ts (hunk 15 above)
Comment (cvarjao) [reply test/commands/git/checkout.test.ts:3]: Agreed. Also sounds like a good contribution to this PR to validate the understanding and familiarize with the tools/frameworks.

File: src/api/model/vault/entry.ts (hunk 1 above)
Comment (cvarjao) [reply src/api/model/vault/entry.ts:1]: Probably. But this vault thing is early proposal. We need more conversation. Consider this an internal API atm.

File: src/api/service/axios-factory.ts (hunk 3 above)
Comment (cvarjao) [reply src/api/service/axios-factory.ts:13]: Updated to read from environment variable and default to those values.

File: src/api/service/axios-factory.ts (hunk 4 above)
Comment (cvarjao) [reply src/api/service/axios-factory.ts]: moved to a function

File: README.md [hunk 17]
```diff
@@ -24,7 +24,7 @@ $ npm install -g @bcgov/nrdk
 $ nrdk COMMAND
 running command...
 $ nrdk (-v|--version|version)
-@bcgov/nrdk/0.1.0-rc.28 darwin-x64 node-v12.14.1
+@bcgov/nrdk/0.1.0-rc.39 darwin-x64 node-v12.14.1
```
Comment (2011sandeepraj) [inline README.md:27]: Can you please add description to preferably all 3 commands nrdk clean, deploy and on:jira:issue but atleast to the last one

**Review by 2011sandeepraj — APPROVED** [review]
Nothing that could catch the eye.

File: README.md (hunk 17 above)
Comment (cvarjao) [reply README.md:27]: the README file is auto generated/updated by oclif
